### ISSUE TITLE
Enhance pydantic types, add info about actual expected type of a value.

### ISSUE THREAD

**Issue by Bobronium** [pydantic/pydantic#1055]
# Feature Request
As per https://github.com/koxudaxi/pydantic-pycharm-plugin/issues/36, it's hard to PyCharm plugin to reveal actual expected type for the field.

Same goes for  #1034 (https://github.com/samuelcolvin/pydantic/pull/1034#discussion_r352244262)

Why wouldn't we have special attribute, pointing to the actual expected type on all pydantic specific types?

Maybe have a base-generic class and declare types by inheritance from it, like:

```py
class PydanticType(Generic[T]): 
    __expected_type__ = T
    __get_validators__: Callable[..., CallableGenerator]
    ...

# and then:
class StrictBool(bool, PydanticType[bool]:
    ...
```

**Comment by samuelcolvin** [comment]
I understand the problem, but what do we do about built in types? Although there aren't that many of them, they're the vast majority of actual usage.

Instead we could have a big dictionary (or function) for lookup up types to show input types are permitted.

Even this wouldn't work for all cases, e.g. `'123'` is a valid input for type int, but `'xxx'` is not.

**Comment by dmontagu** [comment]
I wouldn't mind if the default behavior was to assume the `__expected_type__` was the type itself, and only set anything on other types when that is not the case.

I believe it *should* be possible to statically look for such information while building the pydantic `__init__` signature (in mypy / pycharm). Though it might not be very fun to implement.

I wouldn't mind if `int` allowed `str` as inputs in `__init__` (at least, in non-strict mode), even though it could sometimes fail. It would still be nice though for `StrictInt` to raise a mypy error if I passed a `str` for that field.

**Comment by MrMrRobat** [comment]
I actually don't get the point about built-in types. 
I was able to implement it similar way as described in original post, nothing broke yet. Am I missing something?

Current API:
```py
class StrictStr(ConstrainedStr, PydanticType[str]):
    ....
```
You can see implementation details here: 
https://github.com/samuelcolvin/pydantic/compare/master...MrMrRobat:pydantic-types

**Comment by ycd** [comment]
Any updates on this?

**Comment by samuelcolvin** [comment]
fixed in #5324, some more work to do.
